Bio/PDB/binary_cif.py

```diff
@@ -0,0 +1,301 @@
+    "IntervalQuantization": _interval_quantization_decoder,
+    "RunLength": _run_length_decoder,
+    "Delta": _delta_decoder,
+    "IntegerPacking": _integer_packing_decoder,
+    "StringArray": _string_array_decoder,
+}
+
+
+def _decode(column):
+    # Note that decode modifies the column.
+    encodings = deque(column["data"]["encoding"])
+    column["data"]["encoding"] = encodings
+
+    while encodings:
+        encoding = encodings[-1]
+        _decoders[encoding["kind"]](column)
+
+    return column["data"]["data"]
+
+
+class BinaryCIFParser:
+    """A parser for BinaryCIF files.
+
+    See the `BinaryCIF specification <https://github.com/molstar/BinaryCIF>`_.
+    """
+
+    def __init__(self):
+        """Initialize a BinaryCIF parser."""
+        self._structure_builder = StructureBuilder()
+
+    def _get_hetero_field(self, atom_group: str, component_id: str) -> str:
+        if atom_group == "HETATM":
+            hetero_field = "W" if component_id in ("HOH", "WAT") else "H"
+        else:
+            hetero_field = " "
+
+        return hetero_field
+
+    def _get_residue_ids(self, columns):
+        atom_groups = _decode(columns["_atom_site.group_PDB"])
+        component_ids = _decode(columns["_atom_site.label_comp_id"])
+        hetero_fields = [
+            self._get_hetero_field(atom_group, component_id)
+            for atom_group, component_id in zip(atom_groups, component_ids)
+        ]
+        insertion_codes = [
+            code or " " for code in _decode(columns["_atom_site.pdbx_PDB_ins_code"])
+        ]
+        sequence_ids = _decode(columns["_atom_site.auth_seq_id"])
+
+        return list(zip(hetero_fields, sequence_ids, insertion_codes))
+
+    def _get_atoms(self, columns):
+        names = _decode(columns["_atom_site.label_atom_id"])
+        x_list = _decode(columns["_atom_site.Cartn_x"])
+        y_list = _decode(columns["_atom_site.Cartn_y"])
+        z_list = _decode(columns["_atom_site.Cartn_z"])
+        coordinates_list = np.stack((x_list, y_list, z_list), axis=1)
+        b_factors = _decode(columns["_atom_site.B_iso_or_equiv"])
+        occupancies = _decode(columns["_atom_site.occupancy"])
+        alt_ids = [
+            str(alt_id or " ") for alt_id in _decode(columns["_atom_site.label_alt_id"])
+        ]
+        serial_numbers = _decode(columns["_atom_site.id"])
+        type_symbols = _decode(columns["_atom_site.type_symbol"])
+
+        return [
+            {
+                "name": names[index],
+                "coord": coordinates_list[index],
+                "b_factor": b_factors[index],
+                "occupancy": occupancies[index],
+                "altloc": alt_ids[index],
+                "fullname": names[index],
+                "serial_number": serial_numbers[index],
+                "element": type_symbols[index],
+            }
+            for index in range(len(serial_numbers))
+        ]
+
+    def get_structure(self, source: str) -> Structure:
```

Comment:
As much as I don't like it, I'd keep the signature similar to the other parsers, e.g. add the id argument:

```python
def get_structure(self, id: str, source: str) -> Structure:
    ....
```